File: 1586-longest-subarray-of-1s-after-deleting-one-element/1586-longest-subarray-of-1s-after-deleting-one-element.cpp
// class Solution {
// public:
//     int longestSubarray(vector<int>& nums) {
//         int zeroCount = 0; // Number of zero's in the window.
//         int longestWindow = 0;

//         int l = 0;
//         for(int r = 0; r < nums.size(); r++){
//             zeroCount += (nums[r] == 0);
//             // Shrink the window until the zero counts come under the limit.
//             while(zeroCount > 1){
//                 zeroCount -= (nums[l] == 0);
//                 l++;
//             }

//             longestWindow = max(longestWindow, r-l);
//         }

//         return longestWindow;
//     }
// };

class Solution {
public:
    int longestSubarray(vector<int>& nums) {
        int zeroCount = 1; // Number of zero's to delete.
        int longestWindow = 0;

        int l = 0;
        for(int r = 0; r < nums.size(); r++){
            zeroCount -= (nums[r] == 0);
            // Shrink the window until the zero counts come under the limit.
            while(zeroCount < 0){
                zeroCount += (nums[l] == 0);
                l++;
            }

            longestWindow = max(longestWindow, r-l);
        }

        return longestWindow;
    }
};

File: 1586-longest-subarray-of-1s-after-deleting-one-element/1586-longest-subarray-of-1s-after-deleting-one-element.java
// class Solution {
//     public int longestSubarray(int[] nums) {
//         int l = 0; 
//         int n = nums.length;
//         int maxLength = 0;
//         int zeroCount = 0;

//         for(int r = 0; r < n; r++)
//         {
//             if(nums[r] == 0){
//                 zeroCount++;
//             }

//             while(zeroCount > 1)
//             {
//                 if(nums[l] == 0){
//                     zeroCount--;
//                 }
//                 l++;
//             }
//             maxLength = Math.max(maxLength, r-l);
//         }

//         return maxLength;
//     }
// }

class Solution {
    public int longestSubarray(int[] nums) {
        int l = 0; 
        int n = nums.length;
        int maxLength = 0;
        int zeroCount = 1;

        for(int r = 0; r < n; r++)
        {
            if(nums[r] == 0){
                zeroCount--;
            }

            while(zeroCount < 0)
            {
                if(nums[l] == 0){
                    zeroCount++;
                }
                l++;
            }
            maxLength = Math.max(maxLength, r-l);
        }

        return maxLength;
    }
}